### PR TITLE
treewide/nixos: remove `with lib;` part 1

### PR DESCRIPTION
**Pull request by Stunkymonkey**
## Description of changes

part of #208242

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Stunkymonkey** [comment]
@philiptaron would you consider this "size" of MR ok to review?

I am going to split up the individual services.

(I have already written a shitty script to generate these changes, with a total of 1243 changed files. I will review all files by hand before I submit them via PR.)

**Comment by philiptaron** [comment]
@Stunkymonkey: I think a few things 🙂 

1. I really like one commit per module, named for the module. That helps on many fronts, but _especially_ rebuild and regression hunting.
2. I like about ~150 max commits per PR. That's about my ability to review and say yes. In total, that would be about 8 or 9 PRs for the whole of the `nixos/` directory. Totally manageable. Add me for all of it.
3. I really want to see the tool that you're using, "shitty" or not. That helps me build an intuition for what sort of defects or otherwise to look out for.
4. If there's a module with more than ~50 lines of changes in it, skip that for the bulk PRs.

Thank you for tackling this refactor!

**Comment by Stunkymonkey** [comment]
@philiptaron my "shitty" script (simple replace):

<details><summary>shitty-patch.sh</summary>

```bash
#!/usr/bin/env bash

# Get the list of files containing "with lib;"
files=$(grep -rlFx "with lib;" /home/felix/code/system-installs/nixpkgs/nixos/)

options=(
    "any"
    "attrByPath"
    "attrNames"
    "attrValues"
    "boolToString"
    "cli."
    "concatLists"
    "concatMap"
    "concatStrings"
    "const"
    "elem"
    "escape"
    "filter"
    "flatten"
    "flip"
    "foldAttrs"
    "foldl"
    "foldr"
    "generators."
    "getBin"
    "getDev"
    "getExe"
    "hasAttr"
    "hasPrefix"
    "hasSuffix"
    "imap1"
    "isAttrs"
    "isBool"
    "isDerivation"
    "isInt"
    "isList"
    "isPath"
    "isString"
    "last"
    "length"
    "lessThan"
    "lists"
    "listToAttrs"
    "literalExpression"
    "literalMD"
    "maintainers."
    "makeBinPath"
    "makeLibraryPath"
    "mapAttrs"
    "maybeEnv"
    "mergeOneOption"
    "mkAliasOptionModuleMD"
    "mkBefore"
    "mkDefault"
    "mkEnableOption"
    "mkForce"
    "mkIf"
    "mkImageMediaOverride"
    "mkMerge"
    "mkOption"
    "mkOrder"
    "mkOverride"
    "mkPackageOption"
    "mkRemovedOptionModule"
    "mkRenamedOptionModule"
    "nameValuePair"
    "optional"
    "recursiveUpdate"
    "removePrefix"
    "replaceStrings"
    "setPrio"
    "singleton"
    "sort"
    "stringAfter"
    "stringLength"
    "stringToCharacters"
    "substring"
    "subtractLists"
    "teams."
    "toList"
    "toUpper"
    "types."
    "unique"
    "versionAtLeast"
    "versionOlder"
    "zipAttrs"
    # "all" too many other instances
    # "partition" too many other instances
)

# Function to escape special characters for sed
escape_sed_pattern() {
    echo "$1" | sed 's/[.[\*^$+()|]/\\&/g'
}

# Loop over each file
for file in $files; do
    # delete empty before
    sed -i '/with lib;/,$ b; /^$/d;' "$file"
    # delete empty after
    sed -i '/with lib;/{N;s/\n$//}' "$file"
    # delete match
    sed -i '/^\s*with lib;\s*$/d' "$file"

    for str in "${options[@]}"; do
        # Escape the string for use in sed
        escaped_str=$(escape_sed_pattern "$str")

        # Prepend "lib." to the current string
        lib_str="lib.$str"

        # Escape the lib_str for use in sed
        escaped_lib_str=$(escape_sed_pattern "$lib_str")

        # Use sed to replace occurrences in the temporary file
        sed -i '/inherit/!s/'"$escaped_str"'/'"$escaped_lib_str"'/g' "$file"
    done

    sed -i "s/with maintainers;/with lib.maintainers;/g" "$file"
    sed -i "s/with types;/with lib.types;/g" "$file"
    # revert wrong locations
    sed -i "s/mlib.any/many/g" "$file"
done

```

</details>

> ~150 max commits per PR

I guess you wanted to say about "lines of changes"? 

I started with #336404

**Comment by philiptaron** [comment]
No, I'm saying 150 commits (150 files) 😁 

I feel quite adept at reviewing patterns.

**Comment by philiptaron** [comment]
The tool sadly makes a few bugs -- see if you can evolve it a bit so it doesn't make this:

https://github.com/NixOS/nixpkgs/blob/c8503c7029fcbbac6a0f190b9580b6067e217c8b/nixos/modules/security/pam.nix#L1607-L1608

File: nixos/modules/services/logging/logcheck.nix [hunk 1]
```diff
@@ -43,19 +40,19 @@ let
       };
 
   writeIgnoreCronRule = name: {level, user, regex, cmdline, ...}:
-    let escapeRegex = escape (stringToCharacters "\\[]{}()^$?*+|.");
+    let lib.escapeRegex = lib.escape (lib.stringToCharacters "\\[]{}()^$?*+|.");
```
Comment (philiptaron) [inline nixos/modules/services/logging/logcheck.nix:43]: revert `lib.escapeRegex` or rename binding

File: nixos/modules/services/logging/logcheck.nix [hunk 2]
```diff
@@ -43,19 +40,19 @@ let
...
         cmdline_ = builtins.unsafeDiscardStringContext cmdline;
-        re = if regex != "" then regex else if cmdline_ == "" then ".*" else escapeRegex cmdline_;
+        re = if regex != "" then regex else if cmdline_ == "" then ".*" else lib.escapeRegex cmdline_;
```
Comment (philiptaron) [inline nixos/modules/services/logging/logcheck.nix:45]: revert or rename binding

File: nixos/modules/services/logging/logrotate.nix [hunk 3]
```diff
@@ -1,33 +1,30 @@
...
-with lib;
-
 let
   cfg = config.services.logrotate;
 
   generateLine = n: v:
     if builtins.elem n [ "files" "priority" "enable" "global" ] || v == null then null
     else if builtins.elem n [ "frequency" ] then "${v}\n"
-    else if builtins.elem n [ "firstaction" "lastaction" "prerotate" "postrotate" "preremove" ]
+    else if builtins.elem n [ "firstaction" "lib.lastaction" "prerotate" "postrotate" "preremove" ]
```
Comment (philiptaron) [inline nixos/modules/services/logging/logrotate.nix:8]: revert

File: nixos/modules/services/logging/ulogd.nix [hunk 4]
```diff
@@ -1,16 +1,14 @@
...
-with lib;
 let
   cfg = config.services.ulogd;
   settingsFormat = pkgs.formats.ini { listsAsDuplicateKeys = true; };
   settingsFile = settingsFormat.generate "ulogd.conf" cfg.settings;
 in {
   options = {
     services.ulogd = {
-      enable = mkEnableOption "ulogd, a userspace logging daemon for netfilter/iptables related logging";
+      enable = lib.mkEnableOption "ulogd, a userspace logging daemon for netlib.filter/iptables related logging";
```
Comment (philiptaron) [inline nixos/modules/services/logging/ulogd.nix:9]: revert text change

**Review by philiptaron — CHANGES_REQUESTED** [review]
almost there 💪🏻

File: nixos/modules/config/terminfo.nix [hunk 5]
```diff
@@ -28,10 +25,10 @@ with lib;
   config = {
 
     # can be generated with:
-    # attrNames (filterAttrs
-    #  (_: drv: (builtins.tryEval (isDerivation drv && drv ? terminfo)).value)
+    # lib.attrNames (lib.filterAttrs
+    #  (_: drv: (builtins.tryEval (lib.isDerivation drv && drv ? terminfo)).value)
```
[the remark this answers is not in the document]
Comment (Stunkymonkey) [reply nixos/modules/config/terminfo.nix:29]: IMHO: nothing to do here... right?

[the remark this answers is not in the document]
Comment (philiptaron) [reply nixos/modules/config/terminfo.nix:29]: Looks good to me 🤳🏻

File: nixos/modules/hardware/device-tree.nix [hunk 6]
```diff
@@ -1,40 +1,37 @@
...
+      name = lib.mkOption {
+        type = lib.types.str;
         description = ''
           Name of this overlay
         '';
       };
 
-      filter = mkOption {
-        type = types.nullOr types.str;
+      lib.filter = lib.mkOption {
```
Comment (philiptaron) [inline nixos/modules/hardware/device-tree.nix:14]: `lib.filter =` here is busted.

**Review by philiptaron — CHANGES_REQUESTED** [review]
One more.

**Review by philiptaron — APPROVED** [review]

**Comment by fufexan** [comment]
https://github.com/NixOS/nixpkgs/commit/3b7622d31b50e332a54955d2ef10621ce39ec0e9#diff-9bd6954b6c183d51a0dcbf1f540c61b89a2427e1d3ee7af1dccd98ca2c859279R118 This must've slipped. `utils` has no `lib` inside, and this broke my system rebuild.

**Comment by Aleksanaa** [comment]
How do we currently ensure that the rest of the changes are correct and not just because others haven't used them yet?

**Comment by eclairevoyant** [comment]
easiest way imo would be to fix the script and then rerun it before this PR to identify the breakages. the problem is the missing `\b` at the end which ends up matching substrings.

**Comment by eclairevoyant** [comment]
Okay I've started running it locally, but it seems to be quite slow, not sure when it'll finish.

**Pull request merged**